src/installer/managed/Microsoft.NET.HostModel/AppHost/HostWriter.cs

```diff
@@ -125,28 +130,42 @@ void RewriteAppHost(MemoryMappedFile mappedFile, MemoryMappedViewAccessor access
-
                         // Get the size of the source app host to ensure that we don't write extra data to the destination.
                         // On Windows, the size of the view accessor is rounded up to the next page boundary.
-                        long sourceAppHostLength = appHostSourceStream.Length;
+                        var appHostLength = appHostSourceStream.Length;
+                        memoryMappedFile = MemoryMappedFile.CreateFromFile(appHostSourceStream, null, 0, MemoryMappedFileAccess.Read, HandleInheritability.None, true);
+                        memoryMappedViewAccessor = memoryMappedFile.CreateViewAccessor(0, 0, MemoryMappedFileAccess.CopyOnWrite);
 
                         // Transform the host file in-memory.
                         RewriteAppHost(memoryMappedFile, memoryMappedViewAccessor);
 
-                        // Save the transformed host.
-                        using (FileStream fileStream = new FileStream(appHostDestinationFilePath, FileMode.Create))
+                        MachObjectFile machObjectFile = null;
+                        if (!appHostIsPEImage)
                         {
-                            BinaryUtils.WriteToStream(memoryMappedViewAccessor, fileStream, sourceAppHostLength);
-
-                            // Remove the signature from MachO hosts.
-                            if (!appHostIsPEImage)
+                            if (enableMacOSCodeSign)
+                            {
+                                string fileName = Path.GetFileName(appHostDestinationFilePath);
+                                machObjectFile = new MachObjectFile(memoryMappedViewAccessor, fileName);
+                                appHostLength = machObjectFile.CreateAdHocSignature(memoryMappedViewAccessor, fileName);
+                            }
+                            else
                             {
-                                MachOUtils.RemoveSignature(fileStream);
+                                if (MachObjectFile.TryRemoveCodesign(memoryMappedViewAccessor, out long? length))
+                                {
+                                    appHostLength = length.Value;
+                                }
                             }
+                        }
 
-                            if (assemblyToCopyResourcesFrom != null && appHostIsPEImage)
+                        // Save the transformed host.
+                        using (FileStream fileStream = new FileStream(appHostDestinationFilePath, FileMode.Create, FileAccess.ReadWrite))
+                        {
+                            BinaryUtils.WriteToStream(memoryMappedViewAccessor, fileStream, appHostLength);
+                            if (!appHostIsPEImage && enableMacOSCodeSign)
+                            {
+                                machObjectFile.WriteCodeSignature(fileStream);
```

Comment:
Why split these lines up? That is, why not write the code signature above? Or do all the work to generate the signature here?

Reply:
Since memory mapped files can't be resized, MachObjectFile.CreateAdHocSignature() will write out the headers and load commands to the memory mapped file, but not the signature blob. The signature blob will instead be written to the destination file stream after the rest of the file is copied from the memory mapped file.

Reply:
Oh are files larger after codesigning? I thought the signature blob was already in the file. Could we reserve the space in the apphost template to make signing easier?

Reply:
The platform linker ad-hoc signs the apphost binary when it's built, so there is a signature blob present, but if the file is renamed to a longer name the signature will be slightly larger.

In my first attempt, I tried extending the source apphost and then shrinking it after we write out the destination, but some tests rely on the source apphost being a read-only file, and I think it makes sense to be allow the source to be read-only. I think this also runs into antivirus performance issues. I wasn't able to find a way to make the memory mapped file larger than the file on disk.

We could copy the source apphost to the destination, extend it, and then modify it in place, but my understanding is that modifying the file in place causes antivirus issues.

We could allocate space when we build the apphost binary, but that seems like a significant change just to make this code a bit neater.

For background, the HostWriter uses a `CopyOnWrite` memory mapped accessor to avoid antivirus performance issues, since the modifications to a CopyOnWrite file aren't visible to other processes until they're written to a new file.

Reply:
I looked into CopyOnWrite and there doesn't seem to be a simple way to edit the file in memory and flush all the changes only when the file is closed to avoid the AV issues.